modules/contact/src/ContactSlipDamper.C

```diff
@@ -0,0 +1,226 @@
+                _num_stick_locked++;
+
+              RealVectorValue tangential_inc_slip_prev_iter = info._incremental_slip_prev_iter -
+                                                              (info._incremental_slip_prev_iter * info._normal) * info._normal;
+              RealVectorValue tangential_inc_slip = info._incremental_slip -
+                                                    (info._incremental_slip * info._normal) * info._normal;
+//              RealVectorValue tangential_iterative_slip = tangential_inc_slip - tangential_inc_slip_prev_iter;
+//              RealVectorValue iterative_slip = info._incremental_slip - info._incremental_slip_prev_iter;
+
+              RealVectorValue tangential_it_slip = tangential_inc_slip - tangential_inc_slip_prev_iter;
+              Real node_damping_factor = 1.0;
+              if ((tangential_inc_slip_prev_iter * tangential_inc_slip < 0.0) &&
+                  info._mech_status == PenetrationInfo::MS_SLIPPING_FRICTION)
+              {
+                info._slip_reversed = true;
+                _num_slip_reversed++;
+                Real prev_iter_slip_mag = tangential_inc_slip_prev_iter.norm();
+                RealVectorValue prev_iter_slip_dir = tangential_inc_slip_prev_iter / prev_iter_slip_mag;
+                Real cur_it_slip_in_old_dir = tangential_it_slip * prev_iter_slip_dir;
+
+                if (prev_iter_slip_mag > info._slip_tol ||
+                    cur_it_slip_in_old_dir > -_damping_threshold_factor * prev_iter_slip_mag)
+                  node_damping_factor = 1.0 - (cur_it_slip_in_old_dir + prev_iter_slip_mag) / cur_it_slip_in_old_dir;
+
+                if (node_damping_factor < 0.0)
+                  mooseError("Damping factor can't be negative");
+
+                if (node_damping_factor < _min_damping_factor)
+                  node_damping_factor = _min_damping_factor;
+              }
+
+              if (tangential_it_slip.norm() > _max_iterative_slip)
+                node_damping_factor = (tangential_it_slip.norm() - _max_iterative_slip) / tangential_it_slip.norm();
+
+              if (_debug_output && node_damping_factor < 1.0)
+                _console << "Damping node: " << node->id()
+                  << " prev iter slip: " << info._incremental_slip_prev_iter
+                  << " curr iter slip: "<< info._incremental_slip
+                  << " slip_tol: " << info._slip_tol
+                  << " damping factor: " << node_damping_factor
+                  << "\n";
+
+              if (node_damping_factor < damping)
+                damping = node_damping_factor;
+            }
+          }
+        }
+      }
+    }
+    _console<<std::flush;
+    _communicator.sum(_num_contact_nodes);
+    _communicator.sum(_num_sticking);
+    _communicator.sum(_num_slipping);
+    _communicator.sum(_num_slipping_friction);
+    _communicator.sum(_num_stick_locked);
+    _communicator.sum(_num_slip_reversed);
+    _communicator.min(damping);
```

Comment:
This is fine for now but we can consolidate all these into a single communication later if necessary.